### ISSUE TITLE
iOS: Deadlock in MSAnalytics.TrackEvent causing watchdog transgression

### ISSUE THREAD

**Issue by StefAnglr** [microsoft/appcenter-sdk-dotnet#1047]
### **Description**

We recently updated our App Center SDK nugets from 1.14.0 to 2.1.0 to get the fix for #857.  We use both Crashes and Analytics for tracking crashes, exceptions, and events. We make particularly heavy use of event tracking, logging events every time a view appears or disappears, among many others.

After updating, we started seeing hangs leading to crashes on iOS due to watchdog timeouts (0x8badf00d). We noticed while examining the crash logs that in every log, the main thread was waiting on a lock within an `MSAnalytics.TrackEvent` call, and several other threads were also waiting on locks within `MSAnalytics` or `MSCrashes` methods. After downgrading back to 1.14.0, with no changes to our code, we no longer see these hangs & crashes.

We also have not been seeing crash reports in App Center when using 2.1.0, but I don't know if this issue is related.

### **Repro Steps**
Unfortunately, as is typical with synchronization issues like these, we don't have a consistent reproduction - only crash logs from the cases where it did occur, which are attached.

### **Details**

1. What is your app platform (Xamarin.Android or Xamarin.iOS or UWP)?
    - Xamarin.iOS
2. If using Xamarin.Forms or if using portable/shared code to call our SDK APIs, are you using shared project, PCL code or .NET standard code for the application? Which .NET standard version or which PCL profile?
    - .NET standard 2.0
2. Which SDK version are you using?
    - 2.1.0
4. Which OS version did you experience the issue on?
    - iOS 12.3.1
5. What device version did you see this error on?  Were you using an emulator or a physical device?
    - Physical, iPhone 8
6. What third party libraries are you using? 
    - Xamarin.Auth 1.6.0.2
    - NewtonSoft.Json 12.0.1
    - MvvmCross 6.2.1
    - Nito.AsyncEx 5.0.0
    - AdjustSdk.Xamarin.iOS 4.17.0
    - Esri.ArcGISRuntime 100.4.0
    - FluentMigrator 3.1.3
    - SQLite.NET PCL 1.4.118
7. Please enable verbose logging for your app using `AppCenter.LogLevel = LogLevel.Verbose` before your call to `AppCenter.Start(...)` and include the logs here:
    We have not been able to reproduce this issue with the debugger attached. If these logs would still be useful I can add them.

Crash logs:
[102716.txt](https://github.com/microsoft/appcenter-sdk-dotnet/files/3398315/102716.txt)
[085414.txt](https://github.com/microsoft/appcenter-sdk-dotnet/files/3398317/085414.txt)
[101740.txt](https://github.com/microsoft/appcenter-sdk-dotnet/files/3398321/101740.txt)
[080844.txt](https://github.com/microsoft/appcenter-sdk-dotnet/files/3398324/080844.txt)
[102401.txt](https://github.com/microsoft/appcenter-sdk-dotnet/files/3398326/102401.txt)
[084115.txt](https://github.com/microsoft/appcenter-sdk-dotnet/files/3398327/084115.txt)

**Comment by MatkovIvan** [comment]
Hi @StefAnglr,

Thanks for getting in touch!
Based on the logs, this one has been already fixed by https://github.com/microsoft/appcenter-sdk-apple/pull/1685
It looks like not only `Crashes.TrackError` was affected. Anyway, the fix will be included to July release - `2.2.0-preview` (we're shipping only preview version this month).

Best,
Ivan

**Comment by StefAnglr** [comment]
Thanks for the fix and the quick reply!

**Comment by LeroyStaines** [comment]
I'm having this issue with version 3+

**Comment by MatkovIvan** [comment]
@LeroyStaines most likely it's a separate one. Please open new issue and fill the template (logs etc)
